Classes/ImgproxyBuilder.php

```diff
@@ -44,46 +46,42 @@ public function __construct(string $imgproxyUrl, string $key = null, string $sal
      * @param int $targetHeight
      * @param string $resizingType
      * @param bool $enlarge
-     * @return array{width: int, height: int}
+     * @return Dimension
      * @internal
      * Get the expected size of the resulting image
      */
-    public static function expectedSize(?int $actualWidth, ?int $actualHeight, int $targetWidth, int $targetHeight, string $resizingType, bool $enlarge): array
+    public static function expectedSize(Dimension $actualDimension, Dimension $targetDimension, string $resizingType, bool $enlarge): Dimension
     {
-        $noActualSize = ($actualWidth == null || $actualHeight == null);
-        if ($noActualSize || $resizingType === ImgproxyBuilder::RESIZE_TYPE_FORCE || $resizingType === ImgproxyBuilder::RESIZE_TYPE_FILL) {
-            return [
-                'width' => $targetWidth,
-                'height' => $targetHeight,
-            ];
+        if ($actualDimension->noWidth() || $actualDimension->noHeight()) {
+            return $targetDimension;
         }
 
-        $actualAspectRatio = $actualWidth / $actualHeight;
-        if ($targetWidth === 0 && $targetHeight === 0) {
-            return [
-                'width' => $actualWidth,
-                'height' => $actualHeight,
-            ];
-        } else if ($targetHeight === 0 || $targetWidth === 0) {
-            // Use the actual aspect ratio as the target aspect ratio if one of target width / height is not known
-            $targetAspectRatio = $actualAspectRatio;
-        } else {
-            $targetAspectRatio = $targetWidth / $targetHeight;
+        if ($targetDimension->isZero()) {
+            return $actualDimension;
+        }
+
+        // Use the actual aspect ratio as the default target aspect ratio
+        $targetAspectRatio = $actualDimension->getAspectRatio();
+
+        if (!$targetDimension->noHeight() && !$targetDimension->noWidth()) {
+            $targetAspectRatio = $targetDimension->getAspectRatio();
+        }
+
+        if (!$enlarge && ($targetDimension->isGreater($actualDimension))) {
```

Comment:
I think we need to take the resizing type into account, for `fit` there will be cases where the target dimension is greater in one _dimension_ but it still doesn't need to be enlargened.